### PR TITLE
[2528] Add support for multiple IObjectService

### LINKED ISSUE #2528: Add support for multiple IObjectService
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

With 2023.10.5, only one _IObjectService_ can be provided.
Add a mechanism allowing several _IObjectService_ to live together would be an improvement.

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

No alternatives considered.

### PR DESCRIPTION
**Pull request by AxelRICHARD**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/112_add_support_for_multiple_IObjectService.adoc [hunk 1]
```diff
@@ -0,0 +1,118 @@
+= ADR-112 - Add support for multiple IObjectService
```
Comment (sbegaudeau) [inline doc/adrs/112_add_support_for_multiple_IObjectService.adoc]: - We will definitely move forward with something similar to the second option. There no need for everybody to manage this complexity everywhere.
- I'm not fan of `IObjectServiceProvider` because it does not provide instances of `IObjectService`. I would rather go forward with renaming `ObjectService` to `ComposedObjectService` and introduce an  `IObjectServiceDelegate` which would have to be implemented by those integrating Sirius Components.
- It would be nice to have the default behavior implement some sort of `IDefaultObjectService` (which would be similar to IObjectService). This way, it would allow an application integrating multiple metamodels to define a custom default behavior for every metamodels and small customization where it is necessary in `IObjectServiceDelegate`.
- It should also be noted that the `IObjectService` should/will be separated into smaller interfaces in order to simplify all that (extracting an `ILabelService` and an `IImageService` would be a no brainer). It would also greatly simplify the manipulation of multiple metamodels in a single application and the reusability of metamodels in various Sirius Web applications. No need to copy/paste every time how to retrieve an object from its id if you just want to customize a label.

To sum things up, we need to have an API which would be more similar to the EMF Edit ones without the dependency to EMF.

It should also be noted that this issue was blocking the use of styled strings in the explorer for example.

Comment (AxelRICHARD) [reply doc/adrs/112_add_support_for_multiple_IObjectService.adoc]: Done, except for the separation into smaller interfaces in order to simplify all that. This will be done in a second step.

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/IObjectServiceDelegate.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.Optional;
+
+/**
+ * Interface of the delegation service interacting with domain objects.
+ *
+ * @author arichard
+ */
+public interface IObjectServiceDelegate {
+
+    boolean canHandle(IEditingContext editingContext, Object object);
```
Comment (sbegaudeau) [inline packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/IObjectServiceDelegate.java]: I would not give access to the `IEditingContext` in this method because most methods here do not give us the ability to leverage the editing context and if some application needs to find out the editing context from the object it can be done in application-specific code.

Comment (sbegaudeau) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/IObjectServiceDelegate.java]: After a look at the rest of the code, I'm even more convinced since you have to give `null` most of the time since you can't do anything else. We will never be able to give the editing context in most use cases and applications can retrieve it by themselves if they want so no need for such variable.

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/services/ComposedObjectService.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        this.defaultObjectService = Objects.requireNonNull(defaultObjectService);
+    }
+
+    @Override
+    public String getId(Object object) {
+        return this.objectServiceDelegates.stream()
+                .filter(delegate -> delegate.canHandle(null, object))
+                .findFirst()
+                .map(delegate -> delegate.getId(object))
+                .orElse(this.defaultObjectService.getId(object));
```
Comment (sbegaudeau) [inline packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/services/ComposedObjectService.java]: This is not what we agreed on. In this case, if someone creates a bad delegate returning `null` then the default value would be returned instead and that's exactly what I don't want. If a delegates indicates that it can handle something, it has to handle it properly, no magic fallback. So the code should look like this:


```
var optionalDelegate = this.objectServiceDelegates.stream()
    .filter(delegate -> delegate.canHandle(object))
    .findFirst();
if (optionalDelegate.isPresent()) {
  return optionalDelegate.get().getId(object);
)
return this.defaultObjectService.getId(object);
```

When we will split the `IObjectService` interface into multiple interfaces like `ILabelService` etc, we could simplify the code a bit but it would keep the same philosophy. I fully expect the code to crash and burn as fast as possible if someone tells us that they can handle such an important responsibility and fail to fulfill it.

Comment (sbegaudeau) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/services/ComposedObjectService.java]: Of course, this comment is valid for all the other parts of this code

File: doc/adrs/113_add_support_for_multiple_IObjectService.adoc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+== Decision
+
+* Add new `IObjectServiceDelegate`.
+All Sirius Web applications may provide as many `IObjectServiceDelegate` as they want.
+
+[source,java]
+----
+public interface IObjectServiceDelegate {
+    
+    boolean canHandle(IEditingContext editingContext, Object object);
```
Comment (sbegaudeau) [inline doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: You can also remove the editingContext here (see my comment below) since I didn't catch it at first.

File: doc/adrs/113_add_support_for_multiple_IObjectService.adoc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+----
+
+
+== Status
+
+To be reviewed
+
+== Consequences
+
+No consequences for existing Sirius Web applications.
```
Comment (sbegaudeau) [inline doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: A minor breaking change is that some existing applications may be referencing ObjectService by excluding it and extending it or things like that.

Comment (sbegaudeau) [reply doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: It would be a good place to mention that in the near future, we will split the `IObjectService` interface into smaller interfaces like `IIdentityService`, `ILabelService`, `IContentService` etc. There will be API breaks around this part of the code later.

File: doc/adrs/113_add_support_for_multiple_IObjectService.adoc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      @Override
+      public String getId(Object object) {
+          // Search for a specific provider
+          Optional<IObjectServiceDelegate> optionalObjectServiceProvider = this.objectServiceDelegates.stream().filter(objectServiceProvider -> objectServiceProvider.canHandle(object)).findFirst();
+          if (optionalObjectServiceProvider.isPresent()) {
+            IObjectServiceProvider  objectServiceProvider = optionalObjectServiceProvider.get();
+            return objectServiceProvider.getId(object);
+          } 
+          // Fallback to default behavior
+          return this.defaultObjectService.getId(object);
```
Comment (sbegaudeau) [inline doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: - provider -> delegates
- You are not following this algorithm later :)

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: doc/adrs/113_add_support_for_multiple_IObjectService.adoc (hunk 4 above)
Comment (AxelRICHARD) [reply doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: Done

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/IObjectServiceDelegate.java (hunk 2 above)
Comment (AxelRICHARD) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/IObjectServiceDelegate.java]: Done

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/services/ComposedObjectService.java (hunk 3 above)
Comment (AxelRICHARD) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/services/ComposedObjectService.java]: Done

File: doc/adrs/113_add_support_for_multiple_IObjectService.adoc (hunk 5 above)
Comment (AxelRICHARD) [reply doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: Done

File: doc/adrs/113_add_support_for_multiple_IObjectService.adoc (hunk 6 above)
Comment (AxelRICHARD) [reply doc/adrs/113_add_support_for_multiple_IObjectService.adoc]: Done

**Review by sbegaudeau — APPROVED** [review]
The code looks good, I have not executed it myself but I believe that everything would crash and burn in case of an error in this part of the code so I'll merge it :)
